### PR TITLE
The revised OIFS Scheme

### PR DESCRIPTION
**Pull request by Kamyarmvn**
This pull request is the revised version of the OIFS scheme discussed in PR #1306. 
The code style has been improved to make it more linter friendly.

### CODE REVIEW

**Comment by njansson** [comment]
Very nice! Thanks for the patience, and the hard work on reworking this based on the new submodule structure 🎉

**Review by njansson — APPROVED** [review]

**Comment by Kamyarmvn** [comment]
> Very nice! Thanks for the patience, and the hard work on reworking this based on the new submodule structure 🎉

Thank you very much for your support and patience during this PR. 
There is an error on precon_fctry. Do you have any suggestions on how to fix it? It seems to be new.

**Comment by njansson** [comment]
> > Very nice! Thanks for the patience, and the hard work on reworking this based on the new submodule structure 🎉
> 
> Thank you very much for your support and patience during this PR. There is an error on precon_fctry. Do you have any suggestions on how to fix it? It seems to be new.

These build errors can be ignored, it will still be merged once a second review is added. The longer story is that NVIDIA Fortran unfortunately doesn't support the entire F2008 standard, thus it struggles with some submodule constructs. We keep the Nvidia build in the ci pipeline to keep track of things

File: src/fluid/bcknd/advection/adv_oifs.f90 [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    this%wlag => wlag
+    this%ctlag => tlag
+    this%dctlag => dtlag
+    this%oifs_scheme => time_scheme
+
+    ! Initializing the convecting fields
+    ! Map the velocity fields from GLL space to GL space
+    call this%GLL_to_GL%map(this%cx, this%ulag%f%x, nel, this%Xh_GL)
+    call this%GLL_to_GL%map(this%cy, this%vlag%f%x, nel, this%Xh_GL)
+    call this%GLL_to_GL%map(this%cz, this%wlag%f%x, nel, this%Xh_GL)
```
Comment (timfelle) [inline src/fluid/bcknd/advection/adv_oifs.f90:176]: I get an error here when attempting to run tgv with oifs on CUDA devices.

```
 *** ERROR: Array not associated with device ***
ERROR STOP 

Error termination. Backtrace:
#0  0x7f08aaf1d960 in ???
#1  0x7f08aaf1e4d9 in ???
#2  0x7f08aaf1fb47 in ???
#3  0x55c586941c63 in __utils_MOD_neko_error_msg
	at common/utils.f90:194
#4  0x55c58691cfe0 in __device_MOD_device_get_ptr_r2
	at device/device.F90:958
#5  0x55c5868f6926 in __tensor_MOD_tnsr3d
	at math/tensor.f90:234
#6  0x55c586958e5e in __interpolation_MOD_interpolator_map
	at sem/interpolation.f90:178
#7  0x55c5869ff5b0 in __adv_oifs_MOD_adv_oifs_init
	at fluid/bcknd/advection/adv_oifs.f90:174
#8  0x55c5869f80bf in __advection_MOD_advection_factory
	at fluid/advection_fctry.f90:108
#9  0x55c5869f2c63 in __fluid_pnpn_MOD_fluid_pnpn_init
	at fluid/fluid_pnpn.f90:360
#10  0x55c5869d6514 in case_init_common
	at /home/tife/Projects/neko-top/external/neko/src/case.f90:230
#11  0x55c5869d8feb in __case_MOD_case_init_from_file
	at /home/tife/Projects/neko-top/external/neko/src/case.f90:120
#12  0x55c586932354 in __neko_MOD_neko_init
	at /home/tife/Projects/neko-top/external/neko/src/neko.f90:284
#13  0x55c5868d7274 in usrneko
	at /home/tife/Projects/neko-top/logs/neko/tgv/usr_driver.f90:7
#14  0x55c5868d1d0e in main
	at /home/tife/Projects/neko-top/logs/neko/tgv/usr_driver.f90:2
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[58102,1],0]
  Exit code:    1
--------------------------------------------------------------------------

```

Comment (njansson) [reply src/fluid/bcknd/advection/adv_oifs.f90:176]: This PR should only add CPU/SX support afaik

**Pull request merged**